goLint: add a check that go.mod is tidied

marten-seemann opened the pull request:
<!---
Please provide answers in the spaces below each prompt, where applicable.
Not every PR requires responses for each prompt.
Use your discretion.
-->
#### Name of feature:

Check that `go.mod` is tidied.

#### Pain or issue this feature alleviates:

There's no reason to commit a `go.mod` that is not tidied.

#### Why is this important to the project (if not answered above):

Best case an untidy `go.mod` pulls in dependencies that are not needed, worst case it leads to build failures due to incorrect module hashes.

#### Is there documentation on how to use this feature? If so, where?

https://go.dev/ref/mod#go-mod-tidy

#### In what environments or workflows is this feature supported?

All of them.

#### In what environments or workflows is this feature explicitly NOT supported (if any)?

None.

#### Supporting links/other PRs/issues:

I've been [running this logic in quic-go](https://github.com/quic-go/quic-go/blob/v0.45.1/.github/workflows/lint.yml#L21-L28)  for a long time.

💔Thank you!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/smallstep/workflows?pullRequest=102) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/smallstep/workflows?pullRequest=102) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/smallstep/workflows?pullRequest=102) it.</sub>

dopey commented:
I'm actually worried about this one because recently `go mod tidy` has been adding `gotoolchain` to all our go.mod files, as well as modifying the selected version to be of the format `x.y.z`, rather than `x.y` (which is what we currently use). We should probably agree to this change as a team before merging, as this check may start failing for a large proportion of our repos.

dopey reviewed (changes requested):
Worried that this may cause linting errors for most of our repos.

marten-seemann commented:
> as well as modifying the selected version to be of the format `x.y.z`, rather than `x.y` (which is what we currently use).

I don't think that's a valid value for >= Go 1.21. The [documentation](https://go.dev/doc/toolchain) states:
> The standard Go toolchains are named goV where V is a Go version denoting a beta release, release candidate, or release. For example, go1.21rc1 and go1.21.0 are toolchain names; go1.21 and go1.22 are not (the initial releases are go1.21.0 and go1.22.0), but go1.20 and go1.19 are.

---

> I'm actually worried about this one because recently `go mod tidy` has been adding `gotoolchain` to all our go.mod files

It shouldn't add a `gotoolchain` if none is present (I've been running it in quic-go for years, it never once added a toolchain line).

---

> Worried that this may cause linting errors for most of our repos.

I ran `go mod tidy` for the following repos:
* agent
* autocert
* bouncer
* certificates
* cli
* cli-utils
* crypto
* guardian
* inventory
* step-issuer

In none of these repos did it introduce any changes.

marten-seemann commented:
Coming back to this after a long time, since `go mod tidy` it has bitten me again (this time in https://github.com/smallstep/relay-load-balancer/pull/30#issuecomment-2803750992).

I updated this PR, and introduced a config flag: `lint-skip-go-mod-tidy`. This flag defaults to `true`, so merging this PR won't break any existing repos. Repos can opt into this check by setting this input variable.

If we desire to make this a standard check, we can first enable it in our repos, one by one, and at some point flip the default value. If we don't want to make it a standard check, that's ok as well: Repos that wish to perform this check can enable it (and that's what I intend to do for both relay and relay load balancer).

marten-seemann commented on .github/workflows/goLint.yml:
```diff
@@ -66,8 +70,17 @@ jobs:
           git config --global url.https://${{ secrets.PAT }}@github.com/.insteadOf git+ssh://git@github.com
           git config --global url.git@github.com:.insteadOf https://github.com/
+      - name: Check that go.mod is tidied
+        if: inputs.skip-go-mod-tidy == false
```
I'm not sure why this needs to be treated as a string, but actionlint complains if I use a ! to invert the boolean.

marten-seemann commented:
It works: 
* Successful execution: https://github.com/smallstep/relay/actions/runs/14462869445/job/40558623156?pr=88
* This is how it looks like when go.mod is not properly tidied: https://github.com/smallstep/relay/actions/runs/14462897313/job/40558706454

hslatman commented:
> It works:
> 
>     * Successful execution: https://github.com/smallstep/relay/actions/runs/14462869445/job/40558623156?pr=88
> 
>     * This is how it looks like when go.mod is not properly tidied: https://github.com/smallstep/relay/actions/runs/14462897313/job/40558706454

Here's the failure result: https://github.com/smallstep/relay/actions/runs/14462897313/job/40558706454#step:8:47. The other is for the generated files.

Looks OK, though.
